cmd/snap/cmd_connections.go

```diff
@@ -0,0 +1,222 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+
+	"github.com/jessevdk/go-flags"
+)
+
+type cmdConnections struct {
+	clientMixin
+	All          bool `long:"all"`
+	Disconnected bool `long:"disconnected"`
+	Positionals  struct {
+		Snap installedSnapName `skip-help:"true"`
+	} `positional-args:"true"`
+}
+
+var shortConnectionsHelp = i18n.G("List interface connections")
+var longConnectionsHelp = i18n.G(`
+List connections between plugs and slots. When passed an optional
+snap name, lists connected and unconnected plugs and slots for
+that snap only.
+
+Pass --all to list connected and unconnected plugs and slots.
+`)
+
+func init() {
+	addCommand("connections", shortConnectionsHelp, longConnectionsHelp, func() flags.Commander {
+		return &cmdConnections{}
+	}, map[string]string{
+		"all":          i18n.G("Show connected and unconnected plugs and slots"),
+		"disconnected": i18n.G("Show disconnected plugs and slots"),
+	}, []argDesc{{
+		// TRANSLATORS: This needs to be wrapped in <>s.
+		name: i18n.G("<snap>"),
+		// TRANSLATORS: This should not start with a lowercase letter.
+		desc: i18n.G("Constrain listing to a specific snap"),
+	}})
+}
+
+func isSystemSnap(snap string) bool {
+	return snap == "core" || snap == "snapd" || snap == "system"
+}
+
+func endpoint(snap, name string) string {
+	if isSystemSnap(snap) {
+		return ":" + name
+	}
+	return snap + ":" + name
+}
+
+func wantedSnapMatches(name, wanted string) bool {
+	if wanted == "system" {
+		switch name {
+		case "core", "snapd", "system":
+			return true
+		default:
+			return false
+		}
+	}
+	return wanted == name
+}
+
+type connectionNotes struct {
+	slot      string
+	plug      string
+	manual    bool
+	gadget    bool
+	undesired bool
+}
+
+func (cn connectionNotes) String() string {
+	opts := []string{}
+	if cn.undesired {
+		opts = append(opts, "disconnected")
+	}
+	if cn.manual {
+		opts = append(opts, "manual")
+	}
+	if cn.gadget {
+		opts = append(opts, "gadget")
+	}
+	if len(opts) == 0 {
+		return "-"
+	}
+	return strings.Join(opts, ",")
+}
+
+func connName(conn client.Connection) string {
+	return endpoint(conn.Plug.Snap, conn.Plug.Name) + " " + endpoint(conn.Slot.Snap, conn.Slot.Name)
+}
+
+func (x *cmdConnections) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	opts := client.ConnectionOptions{
+		All: x.All || x.Disconnected,
+	}
+	wanted := string(x.Positionals.Snap)
+	if wanted != "" {
+		// when asking for a single snap, include its disconnected plugs
+		// and slots
+		opts.Snap = wanted
+		opts.All = true
```

Comment:
Given this behaviour, I'd say error out if `--all` is given together with a snap name. Otherwise it's confusing.